### ISSUE TITLE
[css-cascade][css-syntax] New `!revertable` flag to mark a declaration as "can be reverted when IACVT"

### ISSUE THREAD

**Issue by LeaVerou** [w3c/csswg-drafts#10443]
This came out of #5319 but I figured I should make a new post for this for visibility, since it’s a distinct proposal.
#5319 discusses a keyword like `revert-declaration`, however I don’t think a keyword will work, since anything that is part of the value cannot be known in advance (it can be part of a variable for example). 

The problem is that UAs throw the other declarations away when they see a new one in the same scope. It's impossible to know whether the new one will ever use that keyword so they can keep the old ones around too. It works with `revert` and `revert-layer` because the other declarations are kept around *anyway* since they’re in a separate origin/layer.

However, I think a new !-flag may actually work. TBB, but some potential names could be `!revertable`, `!revocable`, `!transient`, `!layered`. Since `!`-flags need to be present at parse time and are always at the end, it doesn’t have the issues that a keyword would.
When a declaration with this flag is encountered, the UA keeps old declarations around instead of replacing them with this.

```css
/* Will not reset border-radius to 0 if --pill-radius is not set, it will just cascade normally */
border-radius: var(--pill-radius) !revertable; 
```

There can be more than one:

```css
* {
	/* Will not override animation if --effect is not set to one of these values */
	animation: if(style(--effect: grow), 1s grow) !revertable;
	animation: if(style(--effect: pulse), 4s pulse infinite) !revertable;
}
```

One open question is how would this work in the CSS OM. But we could just expose the last declaration, and authors can use Typed OM to get all of them (which allows append).

**Comment by andruud** [comment]
There's overlap here with the ideas in https://github.com/w3c/csswg-drafts/issues/1594.

> The problem is that UAs throw the other declarations away when they see a new one in the same scope.

Ideally, we'd just not do that, and make it always possible to revert to the previous "winner" of the cascade (using a keyword). We'd need to investigate the performance implications of this, though.

**Comment by LeaVerou** [comment]
That’s the idea for the opt-in keyword: that authors would not be doing this for every single declaration, so the performance implications should be minimal.

**Comment by Loirooriol** [comment]
```css
#foo {
  height: 1px;
  block-size: 2px;
  height: var(--invalid) !revertable;
}
```

Assuming an horizontal writing mode, should this revert to 2px or 1px? I guess 2px like `revert-layer` would do in different layers, but it seems extra annoying for CSSOM.

I wonder if it would make more sense to add an entirely new rule like

```css
@revertable #foo {
  height: 1px;
  block-size: 2px;
  height: var(--invalid);
}
```

**Comment by LeaVerou** [comment]
> ```css
> #foo {
>   height: 1px;
>   block-size: 2px;
>   height: var(--invalid) !revertable;
> }
> ```
> 
> Assuming an horizontal writing mode, should this revert to 2px or 1px? I guess 2px like `revert-layer` would do in different layers, but it seems extra annoying for CSSOM.
> 
> I wonder if it would make more sense to add an entirely new rule like
> 
> ```css
> @revertable #foo {
>   height: 1px;
>   block-size: 2px;
>   height: var(--invalid);
> }
> ```

This seems orthogonal to the issue we’re discussing? It would revert to whatever this would revert to:

> ```css
> #foo {
>   height: 1px;
>   block-size: 2px;
>   height: var(--invalid) !revertable;
> }
> ```
> 
> Assuming an horizontal writing mode, should this revert to 2px or 1px? I guess 2px like `revert-layer` would do in different layers, but it seems extra annoying for CSSOM.

Not sure how UAs handle the aliasing, but that doesn't seem particularly difficult to define. 

> 
> I wonder if it would make more sense to add an entirely new rule like
> 
> ```css
> @revertable #foo {
>   height: 1px;
>   block-size: 2px;
>   height: yolo;
> }
> ```

I would be *strongly* opposed to doing this as a rule. Rules are far more heavyweight, scoping has to be defined from scratch, and authors cannot port knowledge about it from other parts of CSS. @-rules should be a last resort, not the first thing we reach for because defining syntax with better ergonomics is harder.

**Comment by LeaVerou** [comment]
Agenda+ since I had some discussions with implementers at the CSS WG meeting last week that indicated this is very implementable.

**Comment by Loirooriol** [comment]
> This seems orthogonal to the issue we’re discussing?

It isn't orthogonal. If we are only reverting to the same property, we could just shift down all declarations for the same property to the last one, then store the values together as a list, say that `getPropertyValue` provides the last one, and add `getPropertyValues` to get all of them. Similar for the setter and the priority. But we can at least keep the same indexing as now.

However, if we can revert to other properties, we can't keep assuming that a given property can only appear at most at one index. So it seems the API and the data structures used by browsers will need considerably bigger changes.

> I would be strongly opposed to doing this as a rule. 

Well, I would be strongly opposed to not exposing the necessary information via CSSOM.

And it doesn't seem great to pollute `CSSStyleDeclaration` with several things that aren't needed by normal style rules.

**Comment by LeaVerou** [comment]
Most aliases are known in advance. The only complication with the issue you're pointing out is that the aliasing can be dynamic. These cases are sufficiently few they can just be stored in a data structure (if they aren’t already).

Nothing is polluting `CSSStyleDeclaration`, please read the first post more carefully?
But as a general principle, the ergonomics of the language itself are more important than CSS OM ergonomics. That doesn't mean the CSS OM doesn't matter, but the weightings of tradeoffs are different.

**Comment by Monknow** [comment]
> I wonder if it would make more sense to add an entirely new rule like

> ```css
> @revertable #foo {
>  height: 1px;
>  block-size: 2px;
>  height: var(--invalid);
>}
>```

Even if an at-rule and a `!`-flag are both present after parse time, I think an at-rule would prompt users to put all their styles inside the block, instead of narrowing down which properties they want to enforce the revertable behavior.

Besides, as specified on the original issue at [#5319](https://github.com/w3c/csswg-drafts/issues/5319), the fallback property may be on another CSS rule (or even on third-party styles), like the following:

```html
<p class="lorem">Lorem ipsum dolor</p>
```

```css
p {
 background-color: red; // declaration gets thrown away
} 

.lorem { 
 background-color: var(--not-a-color);  // declaration is IACVT 
}
```

If the rule with the IACVT is more specific, the declaration on the lesser specific rule gets thrown away. If I understand your approach correctly, both declarations would need to be in the at-rule, which is harder if they are in different rules. 

I think putting a `!`-flag (or even the initial proposal for a keyword) after the custom property that may be IACVT is cleaner and gives the user the necessary control.

**Comment by Loirooriol** [comment]
> If I understand your approach correctly, both declarations would need to be in the at-rule

@Monknow No. When the second winner of the cascade is in another rule, then it's no different than `!revertable`. It's not thrown away at parse time, it's just a matter of discussion whether browsers can afford to keep it during the cascade or not.

The problem with normal rules is that duplicate declarations for the same properties get overridden at parse time. So we need a different data structure. Then a different kind of rule may be cleaner.

> Nothing is polluting `CSSStyleDeclaration`, please read the first post more carefully?

@LeaVerou Please read my comment more carefully: "I would be strongly opposed to not exposing the necessary information via CSSOM", which clashes with your proposal in the first post of not exposing it via CSSOM.

**Comment by LeaVerou** [comment]
> > Nothing is polluting `CSSStyleDeclaration`, please read the first post more carefully?
> 
> @LeaVerou Please read my comment more carefully: "I would be strongly opposed to not exposing the necessary information via CSSOM", which clashes with your proposal in the first post of not exposing it via CSSOM.

I did. Exposing it via the CSS OM doesn't mean exposing it via every single CSS OM API. If `.style` does the most commonly needed thing, and `.styleMap` gives you all the information, that sounds like it would meet your requirements?

---

> I would be strongly opposed to doing this as a rule.

That said, there may be value in having an @-rule *as well*, as sugar for applying `!revertable` to a bunch of declarations. But as it was pointed out, it should probably not be done from the start, so that authors don’t use it to wrap all their declarations without thinking. The entire reason `!revertable` might work is because the idea is an !flag is _not_ going to be used everywhere.

**Comment by miragecraft** [comment]
How would the `!revertable` work with custom properties, if at all?

I imagine registered custom property would force variable resolution during type checking, so below should work right?

```css
.example {
  --registered: revert to this value;

  --IACVT: max(invalid);
  --registered: var(--IACVT) !revertable;
}

@property --registered {
  syntax: "<number>";
  inherits: false;
  initial-value:0;
}
```

What about unregistered custom property?

Would the presence of the `!revertable` force the variable to resolve in order to check for IACVT?

```css
.example {
  --unregistered: revert to this value;

  --IACVT: max(invalid);
  --unregistered: var(--IACVT) !revertable;
}
```

**Comment by kizu** [comment]
Good question! For registered custom properties, things should be relatively straightforward and work as for any other custom properties. In your example it will be `0`, as no `--registered` is defined with a valid value, and if we will have something like:

```CSS
.example {
  --registered: 1;
  --IACVT: max(invalid);
  --registered: var(--IACVT) !revertable;
}
```

then it will revert to `1`.

For unregistered custom properties, the most logical thing to do is to treat them as they're treated now: always valid until used, in the same way it currently “taints” any declaration making fallbacks impossible (and for which we'll have `first-valid()` — https://github.com/w3c/csswg-drafts/issues/5055).

I don't know if it is currently possible to determine if an unregistered custom property is invalid _before_ using it? If so, I can see it useful to be able to revert to previous values when doing `!revertable`. Otherwise it will be useless, and making it work this way should not break anything, as `!revertable` is itself new, so we can define how it works for this case (again, if this is technically possible/feasible).

**Comment by miragecraft** [comment]
> In your example it will be `0`, as no `--registered` is defined with a valid value

Good catch on my code snippet, I forgot that the reverted value needs to conform to the syntax specified.

> I don't know if it is currently possible to determine if an unregistered custom property is invalid before using it? If so, I can see it useful to be able to revert to previous values when doing `!revertable`. 

That's what I'm hoping it can do, I believe currently registered property with "any token" syntax behave the same as unregistered property, correct? So simply doing so doesn't solve the problem.

My simplified use case is follows - I want to create a design system where you can specify width as fractions.

```css
.example {
  width: calc(100% * var(--width));
  --width: 2/3;
  --width: max(invalid) !revertable;
}
```

As you can see here there isn't any syntax I can use other than any token, since "2/3" doesn't conform to any type.

Of course I can use `calc(2/3)` instead, but it's no longer elegantly expressed.

If we can force IACVT checking with `!revertable` then this problem would be solved.

**Comment by LeaVerou** [comment]
The thing is, `--width: max(invalid)` is_not_ IACVT — it just makes other declarations IACVT. `!revertable` can't do anything if the property is not OACVT.

**Comment by miragecraft** [comment]
> The thing is, `--width: max(invalid)` is _not_ IACVT — it just makes other declarations IACVT. `!revertable` can't do anything if the property is not IACVT.

I think this use case is more of a limitation of `@property` than `!revertable`, having `!revertable` work for registered custom property (that's not using `syntax:"*"`) is good enough.

**Comment by Loirooriol** [comment]
Yes, we can't decide whether a `--width` declaration is valid or not depending on where `var(--width)` is used, because it could be used in different places which expect different grammars.
You would need to define it with `syntax: "<number> | <number> / <number>"`, which I don't think is valid.

**Comment by dbaron** [comment]
Both https://en.wiktionary.org/wiki/revertible and [Google books ngrams search](https://books.google.com/ngrams/graph?content=revertible%2C+revertable&year_start=1800&year_end=2022&corpus=en&smoothing=3) suggest that "revertible" is the more common spelling than "revertable".

(Though I'm a little hesitant about the name because I think people will have trouble spelling it either way.)

**Comment by benface** [comment]
I apologize if this was answered in the minutes, I had some trouble understanding all of it... but why `revert-rule` and not `revert-declaration`? Does it actually revert all the declarations in the rule as soon as one property is set to that value? If so, that sounds really surprising.

**Comment by benface** [comment]
Ah, I just thought about it and can see why `revert-rule` makes sense. It reverts the property's value to whatever the previous _rule_ that matches sets it to, kinda like how `revert-layer` reverts it to what the @layer below the current one sets it to.

Still, I wonder if a simpler keyword like `cancel` would make more sense, because it's basically used to cancel the current declaration, right?

**Comment by Loirooriol** [comment]
No, it doesn't just cancel the current declaration. It reverts to another rule, so it can also cancel other declarations in the current rule.

**Comment by benface** [comment]
Ah, so the name is not actually counterintuitive, but the feature is, IMO. 😅 What's the reason for giving any single declaration the ability to affect all the declarations of the entire rule it's in? Seems like a first in CSS. Does it include declarations in parent rules, in the case of nesting?

**Comment by benface** [comment]
So if I'm understanding it right, in this example:
```css
button {
  background-color: red;
  --maybe-revert: revert-rule;
}

button:hover {
  --maybe-revert: dont-actually-revert;
}
```
The button's background color would be red on hover only?

**Comment by Loirooriol** [comment]
See my comments and meeting minutes where I explained why just cancelling the current declaration shouldn't be done with style rules (it would need to be a new kind of rule).

Also, I didn't say it cancels ALL declarations in the current rule. It cancels declarations for the same property, or for properties that share a computed value.

**Comment by benface** [comment]
Ahhhh, my bad. Thank you @Loirooriol, I understand now.

**Comment by yisibl** [comment]
Blink implements the `revert-rule` here: https://chromium-review.googlesource.com/c/chromium/src/+/6884901

**Comment by cdoublev** [comment]
@andruud, did you intentionally omit to accept `revert-rule` in [`all`](https://drafts.csswg.org/css-cascade-5/#all-shorthand) property? or is it simply an oversight?

  > **Name:** `all`
  > **Value:** `initial | inherit | unset | revert | revert-layer`

**Comment by andruud** [comment]
@cdoublev Thanks, that's just an oversight. Will fix.

**Comment by yisibl** [comment]
I believe the specification should explicitly state that `revert-rule` does not support reverting within a “declaration block.”
I only discovered this after carefully reviewing the meeting minutes, and it is easily misunderstood.

If I'm not mistaken, in the following style, the color ultimately **isn't green**.

```css
.foo {
  color: green;
  color: var(--unknow, revert-rule); /* ❌ Cannot revert to green */
}
```

**Comment by yisibl** [comment]
Additionally, can CSS preprocessors safely perform polyfills? See https://github.com/parcel-bundler/lightningcss/issues/580#issuecomment-3815797649

**Comment by kizu** [comment]
@yisibl Yep, it is a “gotcha” based on how we serialize things, easily workaroundable as

```
.foo {
  color: green;
  & {
    color: var(--unknow, revert-rule);
  }
}
```

It will benefit from a note & example in the spec for sure.

- - - 

A really great point about preprocessors! They will need to learn not to remove/combine this for sure when minifying it. Some minifiers, for example, could keep both declarations but combine them into one rule, which will be incorrect.

That is workaroundable on the author's side by modifying one of the rules to be _different_, like

```CSS
.test {
  font-size: 36px;
  color: green;
}
.test:where(*) {
  color: var(--foo, revert-rule);
}
```

But it is something to watch out for.

**Comment by yisibl** [comment]
> Yep, it is a “gotcha” based on how we serialize things, easily workaroundable as
> 
> ```
> .foo {
>   color: green;
>   & {
>     color: var(--unknow, revert-rule);
>   }
> }
> ```
> 
> It will benefit from a note & example in the spec for sure.

Adding an empty `& {}` also makes `color: green` take effect(Chrome 147.0.7715.0). I think this part of the specification should include some clarification. cc @andruud 

```html
data:text/html;charset=UTF-8,<!doctype html>
<style>
.foo {
  color: green;
  & {}
  color: var(--unknow, revert-rule);
}
</style>
<div class="foo">Test</div>
```
